sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -209,6 +209,18 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       Row(2743272264L, 2180413220L))
   }
 
+  test("misc data_type function") {
+    val df = Seq(("a", false)).toDF("a", "b")
+
+    checkAnswer(
+      df.select(data_type($"a"), data_type($"b")),
```

Comment:
The idea would be to know the type based on the value itself, not by the schema (i.e. the value could be null):
```scala
val df = spark.sparkContext.parallelize(
      StringData(null) ::
      StringData("a") :: Nil).toDF()
df.select(data_type(col("s"))) //you get null and string in this case
df.schema.map(_.dataType.simpleString) // you just get string
```
On the other hand, it'd be nice to have this SQL function as we do in some databases.